### DISCUSSION TITLE
Concept of Macros/Includes in Prisma

### DISCUSSION THREAD

**Discussion by arobinsongit** [prisma/prisma#13645] [Q&A] [answered]
Working on a net new application.  Specifically using CRDB but I don't think that matters for this discussion.

I have created a simple build system using powershell where I order some SQL scripts by name and do replacements with a "macro" concept.

For example:
```
CREATE TABLE content.icon 
(
	--@@Include [{"Include":"IDPrimaryKey"}]
	name STRING(256) NOT NULL,
	svgsource STRING(16384) NULL,
	fillhexcolor string (7) null default '#000000',
	strokehexcolor string (7) null default '#000000',
	--@@Include [{"Include":"SupplementalColumns"}]
	--@@Include [{"Include":"OwnCreateUpdate"}]
);
```
Separately I have a settings object where I have the contents of these "macros".  

```
	includeReplaces = @{
		IDPrimaryKey = "`tid UUID DEFAULT gen_random_uuid() PRIMARY KEY NOT NULL,"
		SupplementalColumns = "`thidden bool not null default false,`n`titemstatus itemstatusEnum default '',"
		OwnCreateUpdate = "`townedBy STRING(256) null,`tcreatedBy STRING(256) null,`n`tcreatedAt timestamptz not null default now(),`n`tupdatedBy STRING(256) null,`n`tupdatedAt timestamptz not null default now(),"
		Templating = "`tisTemplate bool not null default false,"	
```		

The powershell script looks for --@@Include and then does the replace.  Finally it runs cockroach CLI to actually execute the SQL on the DB.  This works really well right now since every time I build I just wipe the schemas and start over.  

When we get closer to production we will be using Prisma for the runtime layer and I'd like to start moving that way for my design/schema layer.  The thing holding me back is I have a LOT of places where I have these includes to promote design consistency but as of yet, I haven't found if Prisma has a way to do this natively?   If necessary I can keep my current build system with my own macro concept and just call Prisma client to do the build instead of CRDB CLI but I don't want to do anything that might

**Comment by austincrim** [comment, marked as answer]
Hey @arobinsongit 👋,

Prisma would not have support for these kind of scripts as written.

Are these mainly one-off scripts or part of your regular application code?

**Reply by arobinsongit** [reply]
Thanks @austincrim.  On the question of one-off, the honest answer is I'm not sure right now.  Right now I'm just building out my schema and exploring how I want to structure my tables.  My primary use case for these includes is to have consistency between tables as well as saving time repeating myself.  For example, right now I call my PK on every table ID.  Maybe a couple weeks from now I'll realize that's dumb and instead I should call it TablePKID (not really) but by using includes I can change it in one place and when I rebuild the DB it's changed everywhere.  Same goes for some collections of columns like CreatedBy/CreatedAt/UpdatedBy/UpdatedAt.   Having a CMS-style system there is goign to be a lot of that.  

I know once I get into production I'll want to take advantage of Migrate and other tooling like that but for pre-production work the includes method is a good scheme that works for me.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
